examples/ui/reactivity.rs

```diff
@@ -0,0 +1,172 @@
+                let interaction = interaction_query.get(trigger.entity()).unwrap();
+                if matches!(interaction, Interaction::Pressed) {
+                    // We can move this value into the closure that we define,
+                    // allowing us to create custom behavior for the button.
+                    let mut counter = counter_query.get_mut(counter_entity).unwrap();
+                    counter.0 += 1;
+                }
+            },
+        );
+}
+
+/// This temporary module prototypes a user-space implementation of the [`OnMutate`] event.
+///
+/// This comes with two key caveats:
+///
+/// 1. Rather than being continually generated on every change between observers,
+/// the list of [`OnMutate`] events is generated once at the start of the frame.
+/// This restricts our ability to react indefinitely within a single frame, but is a good starting point.
+/// 2. [`OnMutate`] will not have a generic parameter: instead, that will be handled via the second [`Trigger`] generic
+/// and a static component ID, like the rest of the lifecycle events. This is just cosmetic.
+///
+/// To make this pattern hold up in practice, we likely need:
+///
+/// 0. Deep integration for the [`OnMutate`] event, so we can check for it in the same way as the other lifecycle events.
+/// 1. Resource equivalents to all of the lifecycle hooks.
+/// 2. Asset equivalents to all of the lifecycle hooks.
+/// 3. Asset change detection.
+///
+/// As follow-up, we definitely want:
+///
+/// 1. Archetype-level change tracking.
+/// 2. A way to automatically detect whether or not change detection triggers are needed.
+/// 3. Better tools to gracefully exit observers when standard operations fail.
+/// 4. Relations to make defining entity-links more robust and simpler.
+/// 5. Nicer picking events to avoid having to use the naive OnMutate<Interaction> pattern.
+///
+/// We might also want:
+///
+/// 1. Syntax sugar to fetch matching components from the triggered entity in observers
+mod on_mutate {
+    use super::*;
+    use std::marker::PhantomData;
+
+    /// A trigger emitted when a component is mutated on an entity.
+    ///
+    /// This must be explicitly generated using [`GenOnMutate::generate_on_mutate`].
+    #[derive(Event, Debug, Clone, Copy)]
+    pub struct OnMutate<C: Component>(PhantomData<C>);
+
+    impl<C: Component> Default for OnMutate<C> {
+        fn default() -> Self {
+            Self(PhantomData)
+        }
+    }
+
+    /// A temporary extension trait used to prototype this functionality.
+    pub trait GenOnMutate {
+        fn generate_on_mutate<C: Component>(&mut self) -> &mut Self;
+    }
+
+    impl GenOnMutate for App {
+        fn generate_on_mutate<C: Component>(&mut self) -> &mut Self {
+            self.add_systems(First, watch_for_mutations::<C>);
+
+            self
+        }
+    }
+
+    fn watch_for_mutations<C: Component>(mut commands: Commands, query: Query<Entity, Changed<C>>) {
```

Comment:
I don't think we should be using / encouraging this approach. Doing this as a normal "single pass over changed" system means that we are driving reactions to completion over the course of (possibly) many frames. The number of frames it will take to resolve a single propagation of changes is significantly expanded by a number of factors:

1. Each mutation that triggers another mutation within a given component type will have a frame of delay if that entity is "behind" this entity in the change query iteration.
2. Each mutation that triggers another mutation across component types will have a frame of delay if the `watch_for_mutations` system of the triggered mutations runs _before_ the `watch_for_mutations` of the originating mutation's type.
3. We only check if a component has changed once per frame, which means if it changes again, that will not be observed until the next frame.

Given that changes propagate in roughly hierarchy order many levels deep across many types, I anticipate resolving a _single_ propagation tree to take _many_ frames (on the order of seconds of clock time).

The "perceived jank cost" of allowing people to do this is too high I think.

Reply:
This would not be a full solve, but if we did these checks in topo-sorted hierarchy order and checked changed components for each relevant type at each entity, that would allow most UI-related changes to propagate in a single frame via a single pass. From there, you could loop multiple times over the hierarchy until all changes have been fully propagated.

Reply:
Quill uses a "run to convergence" strategy: reactions are run in a loop within a single frame, but are required to converge to quiescence within a set number of iterations: https://github.com/viridia/quill/blob/main/crates/bevy_quill_core/src/view.rs#L508

Reply:
Makes sense!